jupyterhub/files/hub/jupyterhub_config.py

```diff
@@ -408,8 +408,13 @@ def camelCaseify(s):
 
     c.JupyterHub.load_roles.append(role)
 
+# respect explicit null command (distinct from unspecified)
+# this avoids relying on KubeSpawner.cmd's default being None
+_unspecified = object()
+specified_cmd = get_config("singleuser.cmd")
+if specified_cmd is not _unspecified:
+    c.Spawner.cmd = specified_cmd
```

Comment:
I don't understand this very well and think there may be a broken assumption here.

With `singleuser.cmd` set at all, to `null` or something else, it will be recognized. In `helm` the values will be merged, but without setting `singleuser: null` or similar can you get rid of `singleuser.cmd` being declared at all.

I think the change of relevance would be to always set config of c.Spawner.cmd, independent of if we are passed a YAML null (Python None) value or not. This means that we will never regonize the kubespawner default value of `cmd`.

```yaml
# both a and b have the `null` value when YAML is parsed below.
a:
b: null
```

Reply:
> I think the change of relevance would be to always set config of c.Spawner.cmd, independent of if we are passed a YAML null (Python None) value or not. This means that we will never regonize the kubespawner default value of cmd.

That's right - it ensures that setting `singleuser.cmd: null` _explicitly_ sets the value to None, rather than _implicitly_ setting it to whatever KubeSpawner has as default (currently: None).

There are 3 ways we can do this:

1. `set_config_if_not_none` (before this PR)
2. `c.Spawner.cmd = get_config("singleuser.cmd")`
3. this PR

These will all result in the same behavior with the _current_ state of everything, specifically that KubeSpawner.cmd's default is None.

Number 1. caused problems because setting `None` is _not_ telling KubeSpawner to use None, it's telling KubeSpawner to use its default, which is _currently_ None but doesn't have to be. By avoiding this, we keep full control of the value in z2jh, which seems important when z2jh's default is not the same as KubeSpawner's default.
This change also means that KubeSpawner's default need not have been changed to allow using the underlying image.

Number 2. and 3. will generally behave the same, because:

1. singeluser.cmd will always be set, unless we change the default values.yaml to omit it for some reason
2. leaving the value unset still results in None, but via KubeSpawner's default instead

If _either_ `values.yaml` changes to leave `singleuser.cmd` actually unspecified, _or_ KubeSpawner changes its default to not be None, then 2. starts to misbehave, whereas 3. is always correct - reflecting explicit values in config.

Note: we have to choose between `cmd: null` meaning 'explicitly choose the unconfigured default' vs 'explicitly choose None', because we can't do both unless we define a special value e.g. `":default:"` to distinguish "don't set anything" from "set it to None". This issue comes up for every config value where None is a valid value and may not be the default (it's _almost_ always the default if it's legal).

Reply:
Yup, that was a typo where it just wouldn't do what I meant it to :)